crates/ethereum-forks/src/hardforks/mod.rs

```diff
@@ -0,0 +1,128 @@
+    fn fork<H: Hardfork>(&self, fork: H) -> ForkCondition;
+
+    /// Get an iterator of all hardforks with their respective activation conditions.
+    fn forks_iter(&self) -> impl Iterator<Item = (&Box<dyn Hardfork + 'static>, ForkCondition)>;
+
+    /// Convenience method to check if a fork is active at a given timestamp.
+    fn is_fork_active_at_timestamp<H: Hardfork>(&self, fork: H, timestamp: u64) -> bool {
+        self.fork(fork).active_at_timestamp(timestamp)
+    }
+
+    /// Convenience method to check if a fork is active at a given block number.
+    fn is_fork_active_at_block<H: Hardfork>(&self, fork: H, block_number: u64) -> bool {
+        self.fork(fork).active_at_block(block_number)
+    }
+}
+
+/// Ordered list of a chain hardforks that implement [`Hardfork`].
+#[derive(Default, Clone, PartialEq, Eq)]
+pub struct ChainHardforks {
+    forks: Vec<(Box<dyn Hardfork>, ForkCondition)>,
+    map: FxHashMap<&'static str, ForkCondition>,
+}
+
+impl ChainHardforks {
+    /// Creates a new [`ChainHardforks`] from a list which **must be ordered** by activation.
+    ///
+    /// Equivalent Ethereum hardforks **must be included** as well.
+    pub fn new(forks: Vec<(Box<dyn Hardfork>, ForkCondition)>) -> Self {
+        let map = forks.iter().map(|(fork, condition)| (fork.name(), *condition)).collect();
+
+        Self { forks, map }
+    }
+
+    /// Total number of hardforks.
+    pub fn len(&self) -> usize {
+        self.forks.len()
+    }
+
+    /// Checks if the fork list is empty.
+    pub fn is_empty(&self) -> bool {
+        self.forks.is_empty()
+    }
+
+    /// Retrieves [`ForkCondition`] from `fork`. If `fork` is not present, returns
+    /// [`ForkCondition::Never`].
+    pub fn fork<H: Hardfork>(&self, fork: H) -> ForkCondition {
+        self.get(fork).unwrap_or(ForkCondition::Never)
+    }
+
+    /// Retrieves [`ForkCondition`] from `fork` if it exists, otherwise `None`.
+    pub fn get<H: Hardfork>(&self, fork: H) -> Option<ForkCondition> {
+        self.map.get(fork.name()).copied()
+    }
+
+    /// Get an iterator of all hardforks with their respective activation conditions.
+    pub fn forks_iter(
+        &self,
+    ) -> impl Iterator<Item = (&Box<dyn Hardfork + 'static>, ForkCondition)> {
+        self.forks.iter().map(|(f, b)| (f, *b))
+    }
+
+    /// Convenience method to check if a fork is active at a given timestamp.
+    pub fn is_fork_active_at_timestamp<H: Hardfork>(&self, fork: H, timestamp: u64) -> bool {
+        self.fork(fork).active_at_timestamp(timestamp)
+    }
+
+    /// Convenience method to check if a fork is active at a given block number.
+    pub fn is_fork_active_at_block<H: Hardfork>(&self, fork: H, block_number: u64) -> bool {
+        self.fork(fork).active_at_block(block_number)
+    }
+
+    /// Inserts `fork` into list, updating with a new [`ForkCondition`] if it already exists.
+    pub fn insert<H: Hardfork>(&mut self, fork: H, condition: ForkCondition) {
```

Comment:
@DaniPopes this requirement currently prevents us from doing `Arc<dyn Hardforks>`

I'm sure we can work around this, but we should do that separately.